rust/chains/hyperlane-sealevel/src/interchain_gas.rs

```diff
@@ -49,12 +79,144 @@ impl InterchainGasPaymaster for SealevelInterchainGasPaymaster {}
 
 /// Struct that retrieves event data for a Sealevel IGP contract
 #[derive(Debug)]
-pub struct SealevelInterchainGasPaymasterIndexer {}
+pub struct SealevelInterchainGasPaymasterIndexer {
+    rpc_client: RpcClientWithDebug,
+    igp: SealevelInterchainGasPaymaster,
+    program_id: Pubkey,
+    sealevel_config: Option<SealevelConf>,
+}
+
+#[derive(Debug, new)]
+pub struct SealevelGasPayment {
+    payment: InterchainGasPayment,
+    log_meta: LogMeta,
+    beneficiary: H256,
+}
 
 impl SealevelInterchainGasPaymasterIndexer {
     /// Create a new Sealevel IGP indexer.
-    pub fn new(_conf: &ConnectionConf, _locator: ContractLocator) -> Self {
-        Self {}
+    pub fn new(
+        conf: &ConnectionConf,
+        locator: ContractLocator,
+        sealevel_config: Option<SealevelConf>,
+    ) -> Self {
+        let program_id = Pubkey::from(<[u8; 32]>::from(locator.address));
+        // Set the `processed` commitment at rpc level
+        let rpc_client = RpcClientWithDebug::new_with_commitment(
+            conf.url.to_string(),
+            CommitmentConfig::processed(),
+        );
+        let igp = SealevelInterchainGasPaymaster::new(conf, locator);
+        Self {
+            program_id,
+            rpc_client,
+            igp,
+            sealevel_config,
+        }
+    }
+
+    async fn get_payment_with_sequence(
+        &self,
+        sequence_number: u64,
+    ) -> ChainResult<SealevelGasPayment> {
+        let payment_bytes = &[
+            &hyperlane_sealevel_igp::accounts::GAS_PAYMENT_DISCRIMINATOR[..],
+            &sequence_number.to_le_bytes()[..],
+        ]
+        .concat();
+        let payment_bytes: String = base64::encode(payment_bytes);
+
+        // First, find all accounts with the matching gas payment data.
+        // To keep responses small in case there is ever more than 1
+        // match, we don't request the full account data, and just request
+        // the `unique_gas_payment_pubkey` field.
+        let memcmp = RpcFilterType::Memcmp(Memcmp {
+            // Ignore the first byte, which is the `initialized` bool flag.
+            offset: 1,
+            bytes: MemcmpEncodedBytes::Base64(payment_bytes),
+            encoding: None,
+        });
+        let config = RpcProgramAccountsConfig {
+            filters: Some(vec![memcmp]),
+            account_config: RpcAccountInfoConfig {
+                encoding: Some(UiAccountEncoding::Base64),
+                // Don't return any data
+                data_slice: Some(UiDataSliceConfig {
+                    offset: 1 + 8 + 8 + 32 + 4 + 32 + 8 + 8, // the offset to get the `unique_gas_payment_pubkey` field
```

Comment:
Uhhh, I am worried about this breaking. If this temporary and if so please add a `todo` and if not, is this actually standard? I thought borsh was supposed to handle this type of thing.

Reply:
Added a todo comment, @tkporter gave me a reason for it at the time, whose mailbox [implementation](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/dce883d3a19b9727196c9d27db06d2e50f60850b/rust/chains/hyperlane-sealevel/src/mailbox.rs#L613) I followed. It may well have been just for the sake of speed and we may be able to determine the offset using borsh instead?